### ISSUE TITLE
modulation by oscillators

### ISSUE THREAD

**Issue by magnetophon** [magnetophon/DigiDrie#15]
You asked for modulation of all parameters by oscillators.

The [osc_modulation branch](https://github.com/magnetophon/DigiDrie/tree/osc_modulation) has that, but in a hacked and buggy way.

It is hacked in the sense that the vector crossfade is gone for now, because it was much easier to do that way.
Instead, you only ever hear osc A.

It's buggy because it sounds like some params are always moving when you mod with an osc, even when all the `macro x->param name` knobs are at the same value.

It *can* produce interesting sounds, but of course you quickly get aliassing.

Please test this branch and let me know if you want me to put more work in it.

**Comment by ryukau** [comment]
Thanks for this experimentation. I'd love to use this modulation, but if it's hard to implement I'm OK with omitting it in `master`.

Currently, envelope -1 output is added to synth output. This is probably one of a buggy part.

Lowpass for each `modulation->FB` would be nice to have. I'd like to control cutoff. Also key following for lowpass cutoff is handy. For filter type, `smooth` may be sufficient. `resonlp` with fixed Q is another option. `lowpass` is probably best suited, but the implementation looks heavy when cutoff is time-varying.

**Comment by magnetophon** [comment]
> Currently, envelope -1 output is added to synth output. This is probably one of a buggy part.

Why do you think that?
As far as I know, it's multiplied by the oscillators, [here](https://github.com/magnetophon/DigiDrie/blob/968b930dcec1b8d8eab3e9de5e2f3e08d4483d2d/lib/DigiDrie.lib#L150).

> Lowpass for each modulation->FB would be nice to have.

Yes, but I'm affraid that would kill the CPU.
The filters we currenty have are also in the path of the FB phase.
I could give the filter of each oscillator separate parameters.
So osc A could be unfiltered, osc B filtered with a moog at a low freq and osc C with an ms20 at a high freq.

That would not cost a lot of CPU, since the needed filters are running anyway.

Not sure if you are aware of this, but in current master, we have:
4 oscillators: A,B,C and D
times 2 for stereo
times 2 for octave cross-fading.

That is 16 oscillators, that each can have a different phase and freq, so the filters on the phase signal need to be separate: there are 16*3 (moog,ms20 and oberheim) = 48 filters running in this synth...

Would you like to have separate filter params for osc A, B, C and D?
There are 2 reasons I didn't do this:
- For a lot of sounds, you want the filter to move the same for all oscs.
- Explosion of number of parameters.

**Comment by ryukau** [comment]
> Why do you think that?

I used [sisco](https://github.com/x42/sisco.lv2), and waveform looked like that.

I tested with following steps:

1. Set one of the `filter->main->mix-><filter tyep>->level` to 1.
2. Set `filter->main->tweak->freq` to 0.
3. Change `envelopes->envelope -1` parameters.

Not sure if this is related, but I reduced `nrEnvelopes` to 2, which is in `DigiDrie.lib`.

> Yes, but I'm affraid that would kill the CPU.

Hmm, I guess I seriously need block diagram around macro/LFO/envelopes. So I can point out "I'd like to put this here" or "I'd like to rewire this to there" and so on. Faust generated diagram helps to some degree, but hard to follow. Without knowing how signals are summing up, I can't say for sure.

Just to make sure, I was thinking about applying a filter to oscillator output that goes to modulation. The number of new filters becomes `4 * nrMacros`.

I said this because I managed to use 128 filters on LightPadSynth. So it's certainly possible.

> That is 16 oscillators, that each can have a different phase and freq, so the filters on the phase signal need to be separate: there are 16*3 (moog,ms20 and oberheim) = 48 filters running in this synth...

Not related to this topic, but 48 filters are quite a lot for a monosynth. I guess my approximation may help a lot here. I'll try integrating it to DigiDrie once I finished the UI.

> Would you like to have separate filter params for osc A, B, C and D?

Yes. But it sounds difficult. I leave the decision to you here.

**Comment by magnetophon** [comment]
> I tested with following steps:

Is the problem also on master, or only on this branch?

> Hmm, I guess I seriously need block diagram around macro/LFO/envelopes.

Maybe I should walk you trough the block diagram by sharing my screen on https://meet.jit.si/DigiDrie ?

>  I guess my approximation may help a lot here.

I hope so!!!

> Yes. But it sounds difficult. I leave the decision to you here.

No, not difficult, it's just a LOT of extra parameters and I'm not sure if I'd want that as a user.
A lot of times, you just want one filter sweeping trough all the oscillators.

Then again, I would really like a way of linking parameters in the gui, and that would solve this "many parameters issue".
We would just link all 4 filter settings by default.

Thoughts?

**Comment by ryukau** [comment]
I'm on https://meet.jit.si/DigiDrie. I'll reply the rest later.

**Comment by ryukau** [comment]
> Is the problem also on master, or only on this branch?

From the chat, perhaps when filter cutoff becomes too low, the phase becomes constant. So that CZ oscillator outputs DC, and envelope is multiplied to that DC.
